spaghetto/src/Spago/Psa.purs

```diff
@@ -0,0 +1,197 @@
+  }
+
+type PsaOptions =
+  { showSource :: Core.ShowSourceCode
+  , stashFile :: Maybe String
+  , censorWarnings :: Boolean
+  , censorLib :: Boolean
+  , censorSrc :: Boolean
+  , censorCodes :: Set ErrorCode
+  , filterCodes :: Set ErrorCode
+  , statVerbosity :: Core.StatVerbosity
+  , strict :: Boolean
+  }
+
+toOutputOptions :: PsaArgs -> PsaOptions -> PsaOutputOptions
+toOutputOptions { libraryDirs, color } options =
+  { color
+  , censorWarnings: options.censorWarnings
+  , censorLib: options.censorLib
+  , censorSrc: options.censorSrc
+  , censorCodes: options.censorCodes
+  , filterCodes: options.filterCodes
+  , statVerbosity: options.statVerbosity
+  , libraryDirs
+  , strict: options.strict
+  }
+
+psaCompile :: forall a. Set.Set FilePath -> Array String -> PsaArgs -> PsaOptions -> Spago (Purs.PursEnv a) Unit
+psaCompile globs pursArgs psaArgs options@{ showSource, stashFile } = do
+  let outputOptions = toOutputOptions psaArgs options
+  stashData <- case stashFile of
+    Just f -> readStashFile f
+    Nothing -> emptyStash
+
+  result <- Purs.compile globs (Array.snoc pursArgs "--json-errors")
+  let
+    result' = case result of
+      Left err -> { output: err.stdout, exitCode: err.exitCode, err: Just err }
+      Right success -> { output: success.stdout, exitCode: Just success.exitCode, err: Nothing }
+  arrErrorsIsEmpty <- forWithIndex (Str.split (Str.Pattern "\n") result'.output) \idx err ->
+    case jsonParser err >>= CA.decode psaResultCodec >>> lmap CA.printJsonDecodeError of
+      Left decodeErrMsg -> do
+        logDebug $ Array.intercalate "\n"
+          [ "Failed to decode PsaResult at index '" <> show idx <> "': " <> decodeErrMsg
+          , "Json was: " <> err
+          ]
+        -- Note to reviewer:
+        -- Should a stash decode error cause a non-zero exit?
+        pure false
+      Right out -> do
+        files <- liftEffect $ Ref.new FO.empty
+        let
+          loadLinesImpl = if showSource == Core.ShowSourceCode then loadLines files else loadNothing
+          filenames = insertFilenames (insertFilenames Set.empty out.errors) out.warnings
+        merged <- mergeWarnings filenames stashData.date stashData.stash out.warnings
+        for_ stashFile \f -> writeStashFile f merged
+
+        out' <- buildOutput loadLinesImpl outputOptions out { warnings = merged }
+
+        liftEffect $ if psaArgs.jsonErrors then JsonPrinter.print out' else DefaultPrinter.print outputOptions out'
+
+        pure $ FO.isEmpty out'.stats.allErrors
+
+  if Array.all identity arrErrorsIsEmpty then do
+    logSuccess "Build succeeded."
+  else do
+    for_ result'.err $ logDebug <<< show
+    die [ "Failed to build." ]
+
+  where
+  insertFilenames = foldr \x s -> maybe s (flip Set.insert s) x.filename
+  loadNothing _ _ = pure Nothing
+
+  isEmptySpan filename pos =
+    filename == "" || pos.startLine == 0 && pos.endLine == 0 && pos.startColumn == 0 && pos.endColumn == 0
+
+  -- TODO: Handle exceptions
```

Comment:
Should we handle these exceptions? 😄

Reply:
Possibly later, but that's what was in the original PSA code :smile: